Upgrade Font Awesome 4 to 6

bennothommo opened the pull request:
Replaces https://github.com/wintercms/winter/pull/445 and https://github.com/wintercms/winter/pull/447.

This PR upgrades Font Awesome 4 to Font Awesome 6, bringing the number of available icons from 675 to 1748, as mentioned in the above PRs. It has been implemented using Font Awesome's LESS files, but keeps everything backwards compatible by shimming the old icon variables and ensuring the old classes point to the correct font weighting and icon.

This does increase the Backend assets from 4.84MB to 5.06MB (an increase of ~220kb) uncompressed - I haven't tested this using GZIP but would estimate it would equate to around 50-60kb, so shouldn't be too much of a hit.

bennothommo committed:
Upgrade Font Awesome 4 to 6

bennothommo committed:
Fix shimmed icons, remove some CSS variables, recompile

bennothommo commented:
### Screenshots

#### Before

![Screenshot 2022-05-06 at 13-54-25 Settings Winter CMS](https://user-images.githubusercontent.com/15900351/167075349-66ae87cb-040f-4f62-becb-8a526ca7ddbf.png)

#### After

![Screenshot 2022-05-06 at 13-54-00 Settings Winter CMS](https://user-images.githubusercontent.com/15900351/167075375-8dd9c38a-4192-4ca2-ae55-f547f1c389d2.png)

LukeTowers commented on modules/system/assets/ui/font/LICENSE.txt:
```diff
@@ -0,0 +1,165 @@
+All brand icons are trademarks of their respective owners. The use of these
+trademarks does not indicate endorsement of the trademark holder by Font
+Awesome, nor vice versa. **Please do not use brand logos for any purpose except
+to represent the company, product, or service to which they refer.**
```
Why do we have this file included twice?

bennothommo replied:
Must've been leftover when I reorganised the new assets. Removed.

LukeTowers commented on modules/system/assets/css/styles.css:
```diff
@@ -368,3176 +368,7744 @@ blockquote.pull-right .small:after{content:'\00A0 \2014'}
+button.close{padding:0;cursor:pointer;background:transparent;border:0;-webkit-appearance:none}
+.icon-0::before,
+.wn-icon-0::before,
+.oc-icon-0::before{content:"\30"}
```
Can we break these icon content definitions into their own compiled file and just reference them as needed so they aren't adding 11k lines to several different asset files? Also, can you verify that the oc- BC prefix is only applied on the old icon classes?

bennothommo replied:
I'm not sure I follow. Do you mean separating the different icon types? We haven't got a mechanism yet that detects when icons are in use, so we can't conditionally load the icon styles. I just followed the same compilation process that the original icons used.

bennothommo replied:
As for the original oc- polyfills, it's only shimming the original LESS variables, but it is applying an oc prefix to all icons, including the new ones. I can limit this to just the original icons if needed.

LukeTowers replied:
@bennothommo I mean that the styles.css and storm.css both had ~10k lines added / changed, would probably be better to load those styles in an icons.css that gets added via the AssetMaker trait methods whenever the its required. If it's not too much bother I'd prefer if the oc- prefixes were only added to the original icon names, that's a lot of duplicated code that is unnecessary and shouldn't be encouraged to be used.

bennothommo replied:
I think it's reasonable to assume that the icons are used *everywhere* in the Backend. Plus, there would be people who are using icons in an ad-hoc way (myself included) who would be forced to add an `addCss` call to their plugins.

My thoughts are this: for 50kb-60kb extra, you're getting nearly 1100 extra icons. It wouldn't be a tall order for a lot of people.

LukeTowers replied:
How about at least we provide it via a CSS import at the top of the file then? Then it's being cached separately and can be shared between the two compiled files.

bennothommo replied:
I've opted to compile the icons into a separate icon CSS file, which is loaded on all Backend pages. I've re-adjusted how the icons are compiled as well so that the base icon styles can be loaded, and allow us to be selective on icons to load, like the exception pages and maintenance or missing database views. This was previously loading all icons (through `modules/system/assets/css/style.css`).

bennothommo committed:
Remove extra license file

LukeTowers commented:
@bennothommo are you also able to update the UI Icon docs in this PR? Should we like to the FontAwesome 6 site instead of hosting it all ourselves?

bennothommo commented:
@LukeTowers I was thinking that, but we do use a different class structure for the icons. They provide a JSON schema of all icons so it may not be too difficult to create a component that can generate the list in our docs.

LukeTowers commented:
16k icons is a **lot**. I'd be fine with just including instructions for how to change their class names to ours and then linking to their docs. They have a custom algolia integration in order to search all of the icons, I don't really want to go down that road on our end too.

bennothommo committed:
Merge branch 'wip/1.2' into wip/1.2-font-awesome-6

bennothommo committed:
Remove explicit links to icon LESS file.

We will load icons independently in another file so it can be cached independently.

bennothommo committed:
Compile icon styling into separate CSS file

bennothommo committed:
Load separate icon file in Backend pages

bennothommo committed:
Move icons into root UI folder.

Also separated the global base styles for icons with the actual icon list, so we can limit the icons loaded (ie. for the standard exception / error views)

bennothommo committed:
Fix font location

bennothommo committed:
Load a subset of icons for the exception / message views

bennothommo committed:
Add docs for icons

WebVPF commented:
@bennothommo there is no `content` property for brand icons in css.
Example: https://github.com/wintercms/winter/blob/wip/1.2/modules/system/assets/ui/icons.css#L6299-L6301
```
.icon-css3:before,
.wn-icon-css3:before,
.oc-icon-css3:before{font-family:'Font Awesome 6 Brands';font-weight:400}
```

![2022-05-12_115701](https://user-images.githubusercontent.com/61043464/168033381-6fd39bd2-7ae8-42b7-a80b-320b77f7293c.jpg)

bennothommo commented:
@WebVPF sorry, it appeared I hadn't included the brand icons in the compiled icon CSS.

They're included now. :)
